bigtable/beam/batch-write-flow-control-example/src/main/java/bigtable/BatchWriteFlowControlExample.java

```diff
@@ -0,0 +1,237 @@
+
+    // SdkHarnessOptions loggingOptions = options.as(SdkHarnessOptions.class);
+    // loggingOptions.getSdkHarnessLogLevelOverrides()
+    //     .addOverrideForPackage(Package.getPackage("com.google.cloud.bigtable.data.v2.stub"),
+    //         LogLevel.DEBUG);
+
+    generateData(options);
+  }
+
+  static void generateData(BigtablePipelineOptions options) {
+
+    System.out.println("Generating " + options.getBigtableRows() + " rows, each "
+        + options.getBigtableColsPerRow() + " columns, " + options.getBigtableBytesPerCol()
+        + " bytes per column, " + options.getBigtableColsPerRow() * options.getBigtableBytesPerCol()
+        + " bytes per row.");
+
+    String generateLabel = String
+        .format("Generate %d rows for table %s", numRows, options.getBigtableTableId());
+    String mutationLabel = String
+        .format("Create mutations that write %d columns of total %d bytes to each row",
+            options.getBigtableColsPerRow(), options.getBigtableColsPerRow()
+                * options.getBigtableBytesPerCol());
+
+    Pipeline p = Pipeline.create(options);
+
+    PCollection<Long> numbers = p.apply(generateLabel, GenerateSequence.from(0).to(numRows));
+
+    if (options.getUseCloudBigtableIo()) {
+      System.out.println("Using CloudBigtableIO");
+      PCollection<org.apache.hadoop.hbase.client.Mutation> mutations = numbers.apply(mutationLabel,
+          ParDo.of(new CreateHbaseMutationFn(options.getBigtableColsPerRow(),
+              options.getBigtableBytesPerCol())));
+
+      mutations.apply(
+          String.format("Write data to table %s via CloudBigtableIO", options.getBigtableTableId()),
+          CloudBigtableIO.writeToTable(new CloudBigtableTableConfiguration.Builder()
+              .withProjectId(options.getProject())
+              .withInstanceId(options.getBigtableInstanceId())
+              .withTableId(options.getBigtableTableId())
+              .withConfiguration(BigtableOptionsFactory.BIGTABLE_ENABLE_BULK_MUTATION_FLOW_CONTROL,
+                  "true")
+              .withConfiguration(BigtableOptionsFactory.BIGTABLE_BULK_MAX_REQUEST_SIZE_BYTES,
+                  "1048576")
+              .build()));
+    } else {
+      System.out.println("Using BigtableIO");
+      PCollection<KV<ByteString, Iterable<Mutation>>>
+          mutations = numbers.apply(mutationLabel,
+          ParDo.of(new CreateMutationFn(options.getBigtableColsPerRow(),
+              options.getBigtableBytesPerCol())));
+
+      BigtableIO.Write write = BigtableIO.write()
+          .withProjectId(options.getProject())
+          .withInstanceId(options.getBigtableInstanceId())
+          .withTableId(options.getBigtableTableId())
+          .withFlowControl(true);  // This enables batch write flow control
+
+      mutations.apply(
+          String.format("Write data to table %s via BigtableIO", options.getBigtableTableId()),
+          write
+      );
+    }
+
+    p.run();
```

Comment:
this block of code is hard to read. Since it is a code sample it should be easy to understand. Please, reformat it so it will look like steps each of the steps calling `apply` method of the pipeline. See the [dataflow-bigquery-read-tablerows](https://cloud.google.com/dataflow/docs/samples/dataflow-bigquery-read-tablerows) sample as a reference.

Reply:
This code sample shows how to enable the flow control feature. It has 2 parts because we have two different connectors and they're configured differently. I was unsure whether it'll worth duplicating the rest of the code to keep the sample simple. Please advice and I can split.

We're going to write the doc that will use this code sample.

Yes I'll add tests.

Reply:
> I was unsure whether it'll worth duplicating the rest of the code to keep the sample simple. Please advice and I can split.

It seems to me that both parts of the `if`/`else` block apply only two transformations where the first transformations are very similar and hard to distinguish in the current version.
I suggest either to look into an option to abstract these transformations (by inheriting from a proper Beam class) or, at least, to implement each of them in separate method with meaningful names.
Another option can be to implement each pipeline in a separate class from the start to the end.

> We're going to write the doc that will use this code sample.

Please either acquire a region tag or create a new tag using devrel/ site and add it to the code.

Reply:
Split into 2 methods. Added region tag.